spec/integration/application/apply_spec.rb

```diff
@@ -540,6 +540,42 @@ def bogus()
     end
   end
 
+  context 'http file sources' do
+    include_context 'https client'
+
+    it "replaces spaces with + path" do
+      Puppet[:server] = '127.0.0.1'
+      query = nil
+
+      response_proc = -> (req, res) {
+        query = req.query
+        res['Content-Type'] = 'text/plain'
+        res.body = "from the server"
+      }
+
+      https = PuppetSpec::HTTPSServer.new
+      https.start_server(response_proc: response_proc) do |https_port|
+        dest = tmpfile('http_file_source')
+
+        manifest = file_containing("manifest.pp", <<~MANIFEST)
+          file { "#{dest}":
+            ensure  => file,
+            source  => "https://#{Puppet[:server]}:#{https_port}/path/to/file?sv=2019-02-02&st=2020-07-28T20:18:53Z&se=2020-07-28T21:03:00Z&sr=b&sp=r&sig=JaZhcqxT4akJcOwUdUGrQB2m1geUoh89iL8WMag8a8c=",
+          }
+        MANIFEST
+
+        expect {
+          apply.command_line.args << manifest
+          apply.run
+          }.to exit_with(0)
+           .and output(%r{Main/File\[#{dest}\]/ensure: defined content as}).to_stdout
+
+        expect(query).to eq("sv"=>"2019-02-02", "st"=>"2020-07-28T20:18:53Z", "se"=>"2020-07-28T21:03:00Z",
+                            "sr"=>"b", "sp"=>"r", "sig"=>"JaZhcqxT4akJcOwUdUGrQB2m1geUoh89iL8WMag8a8c=")
```

Comment:
Maybe better to use `request.query_string` and verify the raw query string is URL escaped as expected.